proposals/trestle-sdk-cli-decoupling.md

```diff
@@ -0,0 +1,160 @@
+
+**Increased Support for SDK Users**: Independent development and testing can lead to enhanced documentation and code quality targeting the SDK user.
+
+## User/User Story (Optional)
+
+None
+
+### Prior discussion and links
+
+Initial discussion occurred in the OSCAL Compass [community meeting](https://www.youtube.com/watch?v=uUGv3HXlTrI&t=19s  ) on June 14, 2024 pertaining to the reusability of `Rule` logic to support this [issue](https://github.com/oscal-compass/compliance-trestle/issues/1475). An [issue](https://github.com/oscal-compass/community/issues/28) in the community repo was subsequently submitted.
+
+## Goals
+
+- `compliance-trestle` continues to contain the `trestle` CLI logic
+- `compliance-trestle-python` is created as the Python SDK repository
+- `compliance-trestle` and other OSCAL Compass projects migration to the new SDK for applicable logic
+
+## Non-Goals
+
+- Fundamentally changing the architecture or design of the SDK or CLI is out of scope.
+- Expanding the SDK to other programming languages is out of scope (future consideration).
+
+## Proposal
+
+The solution proposed in this document is to move certain pieces of the codebase located in the [compliance-trestle](https://github.com/oscal-compass/compliance-trestle) repository to a new repository containing a Python-based SDK for interaction with OSCAL documents.
+
+Below is high-level description of what would be moved:
+
+**OSCAL Classes**: The classes that represet the OSCAL data model.  
+**Core Functionality**: Contains the fundamental APIs for interacting with OSCAL objects.  
+**Abstractions**: Abstractions for common operations like resolution or core OSCAL validation.  
+  
+</details>
+
+## Design Details
+
+Below is diagram with existing `trestle` logic that is proposed to be part of the SDK. Links to where this type of logic exists currently are included:
+
+```mermaid
+classDiagram
+    OSCALBaseModel <|-- Catalog
+    OSCALBaseModel <|-- Profile
+    OSCALBaseModel <|-- ComponentDefinitions
+    OSCALBaseModel <|-- SSP
+    OSCALBaseModel <|-- AssessmentPlan
+    OSCALBaseModel <|-- AssessmentResult
+    OSCALBaseModel <|-- POAM
+    class OSCALBaseModel{
+      +Config
+      +write()
+      +read()
+    }
+    class Catalog{
+      +CatalogAPI
+    }
+    class Validator{
+      +isValid(OSCALBaseModel)
+    }
+    class Resolver{
+      +Fetcher
+      +import()
+      +merge()
+      +modify()
+      +prune()
+    }
+```
+
+### Relevant Links
+
+ - OSCAL Classes: https://github.com/oscal-compass/compliance-trestle/tree/develop/trestle/oscal
+  - Core Logic: https://github.com/oscal-compass/compliance-trestle/blob/develop/trestle/core/models/
+  - Abstractions: 
+    - https://github.com/oscal-compass/compliance-trestle/tree/develop/trestle/core/resolver
+    - https://github.com/oscal-compass/compliance-trestle/blob/dfe892936e5960ad64f6f387dbe5918314049e89/trestle/core/validator.py
+
```

Comment:
`I think there is a miscommunication here. The intention is to include more than the base classes and as @butler54 mentioned above, it provides a place where we can include functionality over time. This is covered in the proposal document. There is an initial diagram and some links to existing trestle packages there. Could you please provide more concrete feedback on the design details? If there is something you believe would be a beneficial addition to the SDK, please add a suggestion or comment there.`

For sure, these classes can be moved to a separate repo. What I question is the value of doing so. It is surely the case that trestle versions are tightly coupled with OSCAL versions today. I'm having trouble imagining that the SDK will not be tightly coupled to OSCAL and that the CLI will not be tightly coupled to the SDK, version-wise. Thus, I'm having trouble understanding the value of separate repo, since a newer version of the SDK might very well be incompatible with and older version of the CLI, and vice-versa.

Reply:
I see an SDK with well-defined interfaces as being useful to not only the CLI, but other projects. This provides a place to create reusable logic that is valuable outside of the context of the CLI. If the SDK has logic to support multiple projects interacting with OSCAL, I would think the CLI would have a different release cycle.

- I agree the SDK would be tightly coupled to the OSCAL version
- I agree the CLI should be coupled with a specific version of the SDK. For reusable functionality, the CLI can interact with the SDK through these defined interfaces (like the `ProfileResolver`).

I do not think the CLI would ever leverage the SDK and support different versions of OSCAL and agree that there would be compatibility issues there. If there is some wording in the proposal that indicates this is a use case, please let me know so I can change it.

This how I imagined the process:

1. Create the release for SDK v1 (OSCAL 1.1.2)
2. CLI at the current major version can migrate to SDK v1
> Future Event - we are ready to move to OSCAL version 1.2.0
3. SDK BREAKING CHANGE - SDK adopts OSCAL 1.2.0 and releases v2 for the CLI and other projects to leverage.
4. CLI BREAKING CHANGE - Move to to SDK v2 to adopt OSCAL 1.2.0

Is this clarification helpful?